Platform/OpenNtfsDxe/Data.c

```diff
@@ -0,0 +1,566 @@
+
+  if (Vcn >= Runlist->NextVcn) {
+    Status = ReadRunListElement (Runlist);
+    if (EFI_ERROR(Status)) {
+      return (UINT64) (-1);
+    }
+
+    return Runlist->CurrentLcn;
+  } else {
+    Delta = Vcn - Runlist->CurrentVcn;
+
+    return Runlist->IsSparse ? 0 : (Runlist->CurrentLcn + Delta);
+  }
+}
+
+STATIC
+EFI_STATUS
+ReadClusters (
+  IN  RUNLIST *Runlist,
+  IN  UINT64  Offset,
+  IN  UINTN   Length,
+  OUT UINT8   *Dest
+  )
+{
+  EFI_STATUS Status;
+  UINT64     Index;
+  UINT64     ClustersTotal;
+  UINT64     Cluster;
+  UINT64     OffsetInsideCluster;
+  UINTN      Size;
+
+  OffsetInsideCluster = Offset & (mClusterSize - 1);
+  Size = mClusterSize;
+  ClustersTotal = DivU64x64Remainder (Length + Offset + mClusterSize - 1, mClusterSize, NULL);
+
+  for (Index = Runlist->TargetVcn; Index < ClustersTotal; Index++) {
+    Cluster = GetLcn (Runlist, Index);
+    if (Cluster == (UINT64)(-1)) {
+      return EFI_DEVICE_ERROR;
+    }
+
+    Cluster = Cluster * mClusterSize;
+
+    if (Index == (ClustersTotal - 1)) {
+      Size = (UINTN) ((Length + Offset) & (mClusterSize - 1));
+
+      if (Size == 0) {
+        Size = mClusterSize;
+      }
+
+      if (Index != Runlist->TargetVcn) {
+        OffsetInsideCluster = 0;
+      }
+    }
+
+    if (Index == Runlist->TargetVcn) {
+      Size -= (UINTN) OffsetInsideCluster;
+    }
+
+    if ((Index != Runlist->TargetVcn) && (Index != (ClustersTotal - 1))) {
+      Size = mClusterSize;
+      OffsetInsideCluster = 0;
+    }
+
+    if (Cluster != 0) {
+      Status = DiskRead (
+        Runlist->Unit.FileSystem,
+        Cluster + OffsetInsideCluster,
+        Size,
+        Dest
+        );
+      if (EFI_ERROR(Status)) {
+        return Status;
+      }
+    } else {
+      SetMem (Dest, Size, 0);
+    }
+
+    Dest += Size;
+  }
+
+  return EFI_SUCCESS;
+}
+
+EFI_STATUS
+EFIAPI
+DiskRead (
+  IN EFI_FS   *FileSystem,
+  IN UINT64   Offset,
+  IN UINTN    Size,
+  IN OUT VOID *Buffer
+  )
+{
+  EFI_STATUS         Status;
+  EFI_BLOCK_IO_MEDIA *Media;
+
+  ASSERT(FileSystem != NULL);
+  ASSERT(FileSystem->DiskIo != NULL);
+  ASSERT(FileSystem->BlockIo != NULL);
+
+  Media = FileSystem->BlockIo->Media;
+
+  Status = FileSystem->DiskIo->ReadDisk (
+    FileSystem->DiskIo,
+    Media->MediaId,
+    Offset,
+    Size,
+    Buffer
+    );
+  if (EFI_ERROR(Status)) {
+    DEBUG((DEBUG_INFO, "NTFS: Could not read disk at address %08x\n", Offset));
+    return Status;
+  }
+
+  return EFI_SUCCESS;
+}
+
+EFI_STATUS
+EFIAPI
+ReadMftRecord (
+  IN  EFI_NTFS_FILE *File,
+  OUT UINT8         *Buffer,
+  IN  UINT64        RecordNumber
+  )
+{
+  EFI_STATUS     Status;
+
+  Status = ReadAttr (
+    &File->MftFile.Attr,
+    Buffer,
+    RecordNumber * mFileRecordSize,
+    mFileRecordSize
+    );
+  if (EFI_ERROR(Status)) {
+    DEBUG((DEBUG_INFO, "NTFS: Could not read MFT Record 0x%Lx\n", RecordNumber));
+    return Status;
+  }
+
+  return Fixup (Buffer, mFileRecordSize, "FILE");
```

Comment:
A better name shall be chosen than `Fixup`.

Reply:
This name follows NTFS documentation.

Reply:
Well, FixupRecord still feels better than just Fixup, but maybe a better name could be thought of.